### PR TITLE
Allows to use a custom resizer instead of the default one configured in the provider.

### LINKED ISSUE #370: Resizer per format
Don't you think that resizer should be assigned to concrete formate in formats section ?

For example I would like to create two thumbnails with different resizer for one media object.

  formats:
    simple: { width: 100, height: 80, resizer: sonata.media.resizer.simple}
    exact_size:   { width: 200, height: 150, resizer: my.bundle.resizer.custom} //first crop and then resize in order to get 200x150 thumbnail

Now I can only assign resizer per all thumbnails what isn't sufficient for my requirements.

### PR DESCRIPTION
**Pull request by nan-guo**
## Subject

Allows to use a custom resizer instead of the default one configured in the provider for each format.

Closes #370 

## Changelog

```markdown
### Added
- ThumbnailCompilerPass.php

### Changed
- FormatThumbnail.php
- AddProviderCompilerPass.php
- Configuration.php
- Configuration.php
- media_context.rst
- advanced_configuration.rst

```

### CODE REVIEW

**Comment by nan-guo** [comment]
I did not understand the build error, please give some advice, thanks.

**Comment by kunicmarko20** [comment]
Seems that the wrong phpunit version is used, cc @sonata-project/contributors did dev-kit change something IIRC?

**Comment by nan-guo** [comment]
@SonataCI please check the configs, thx.

**Comment by greg0ire** [comment]
1. Fetch all new commits: `git fetch --all`. ⬇️ 
2. Rebase on what you fetched interactively: ✂️ 
    1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin repository.
    2. A window will show up with many lines, remove every line but the last ones, which should correspond to your commits.
   2. If you run into a conflict: 💥 
        1. fix it with `git mergetool`. 😷 
        1. continue on your merry way with `git rebase --continue`. ⏩ 
3. Force push to overwrite all this : `git push --force`. ⬆️

File: src/DependencyInjection/SonataMediaExtension.php [hunk 1]
```diff
@@ -15,6 +13,52 @@
 
 use Sonata\ClassificationBundle\Model\CategoryInterface;
 use Sonata\EasyExtendsBundle\Mapper\DoctrineCollector;
+use Sonata\MediaBundle\CDN\CDNInterface;
```
Comment (OskarStark) [inline src/DependencyInjection/SonataMediaExtension.php]: are you sure all these Classes are needed?

Comment (nan-guo) [reply src/DependencyInjection/SonataMediaExtension.php]: I have updated the file

**Comment by nan-guo** [comment]
php 7 feature didn't pass the test php5

File: src/DependencyInjection/Compiler/AddProviderCompilerPass.php [hunk 2]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+declare(strict_types=1);
```
Comment (OskarStark) [inline src/DependencyInjection/Compiler/AddProviderCompilerPass.php]: please remove to make 5.6 pass

File: docs/reference/media_context.rst [hunk 3]
```diff
@@ -8,7 +8,10 @@ features: resize, cdn and database relationship with entities.
 The ``MediaBundle`` tries to solve this situation by introducing ``context``:
 a context has its own set of media providers and its own set of formats.
 That means you can have a ``small`` user picture format and a ``small`` news
-picture format with different sizes and providers. For example:
+picture format with different sizes and providers. 
+Also each format allows to use a custom resizer instead of the default one configured in the provider,
```
Comment (greg0ire) [inline docs/reference/media_context.rst]: Please comply with the sf docs standards:

> [Each line should break approximately after the first word that crosses the 72nd character (so most lines end up being 72-78 characters);](http://symfony.com/doc/current/contributing/documentation/standards.html)

File: src/Thumbnail/FormatThumbnail.php [hunk 4]
```diff
@@ -21,6 +22,11 @@ class FormatThumbnail implements ThumbnailInterface
      */
     private $defaultFormat;
 
+    /**
+     * @var array
```
Comment (greg0ire) [inline src/Thumbnail/FormatThumbnail.php]:
```suggestion
     * @var ResizerInterface[]
```

File: src/Thumbnail/FormatThumbnail.php [hunk 5]
```diff
@@ -29,6 +35,33 @@ public function __construct($defaultFormat)
         $this->defaultFormat = $defaultFormat;
     }
 
+    /**
+     * @param string           $id
+     * @param ResizerInterface $resizer
```
Comment (greg0ire) [inline src/Thumbnail/FormatThumbnail.php]: That one is tautological
```suggestion
```

File: src/Thumbnail/FormatThumbnail.php [hunk 6]
```diff
@@ -29,6 +35,33 @@ public function __construct($defaultFormat)
...
+     * @param string $id
+     *
+     * @throws \Exception
+     *
+     * @return ResizerInterface
+     */
+    public function getResizer($id)
+    {
+        if (!isset($this->resizers[$id])) {
+            throw new \Exception('Resizer with id: '.$id.' is not attached.');
```
Comment (greg0ire) [inline src/Thumbnail/FormatThumbnail.php]: Please quote the id:
```suggestion
            throw new \LogicException(sprintf('Resizer with id: "%s" is not attached.', $id));
```

File: src/Thumbnail/FormatThumbnail.php [hunk 7]
```diff
@@ -79,7 +112,8 @@ public function generate(MediaProviderInterface $provider, MediaInterface $media
         foreach ($provider->getFormats() as $format => $settings) {
             if (substr($format, 0, \strlen($media->getContext())) == $media->getContext() ||
                 MediaProviderInterface::FORMAT_ADMIN === $format) {
-                $provider->getResizer()->resize(
+                $resizer = (isset($settings['resizer']) && ($settings['resizer'])) ? $this->getResizer($settings['resizer']) : $provider->getResizer();
```
Comment (greg0ire) [inline src/Thumbnail/FormatThumbnail.php]:
```suggestion
                $resizer = (isset($settings['resizer']) && ($settings['resizer'])) ? 
                    $this->getResizer($settings['resizer']) : 
                    $provider->getResizer();
```

File: src/Thumbnail/FormatThumbnail.php (hunk 6 above)
Comment (greg0ire) [reply src/Thumbnail/FormatThumbnail.php]: Please add a unit test for this

Comment (OskarStark) [reply src/Thumbnail/FormatThumbnail.php]: And please use sprintf for concatenation

File: src/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -112,6 +112,7 @@ private function addContextsSection(ArrayNodeDefinition $node)
                                         ->scalarNode('quality')->defaultValue(80)->end()
                                         ->scalarNode('format')->defaultValue('jpg')->end()
                                         ->scalarNode('constraint')->defaultValue(true)->end()
+                                        ->scalarNode('resizer')->defaultFalse()->end()
```
Comment (OskarStark) [inline src/DependencyInjection/Configuration.php]: Shouldn‘t website default Null here instead of false @greg0ire ?

Comment (greg0ire) [reply src/DependencyInjection/Configuration.php]: Uuuuh yeah probably, indeed!

File: tests/DependencyInjection/Compiler/AddProviderCompilerPassTest.php [hunk 9]
```diff
@@ -83,6 +83,7 @@ public function testProcess()
                 'quality' => 70,
                 'format' => 'jpg',
                 'constraint' => true,
+                'resizer' => false,
```
Comment (OskarStark) [inline tests/DependencyInjection/Compiler/AddProviderCompilerPassTest.php]: null?

File: src/DependencyInjection/Compiler/AddProviderCompilerPass.php [hunk 10]
```diff
@@ -117,6 +117,7 @@ public function applyFormats(ContainerBuilder $container, array $settings)
                     $config['format'] = isset($config['format']) ? $config['format'] : 'jpg';
                     $config['height'] = isset($config['height']) ? $config['height'] : false;
                     $config['constraint'] = isset($config['constraint']) ? $config['constraint'] : true;
+                    $config['resizer'] = isset($config['resizer']) ? $config['resizer'] : false;
```
Comment (OskarStark) [inline src/DependencyInjection/Compiler/AddProviderCompilerPass.php]: null?

**Comment by OskarStark** [comment]
Please update the CHANGELOG in the PR header and describe what you added, not which files you updated

File: src/DependencyInjection/Compiler/AddProviderCompilerPass.php (hunk 10 above)
Comment (nan-guo) [reply src/DependencyInjection/Compiler/AddProviderCompilerPass.php]: yes, indeed !

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

**Comment by nan-guo** [comment]
@greg0ire it's ok ?

**Comment by OskarStark** [comment]
Looks good to me 👍

**Comment by greg0ire** [comment]
Thanks @nan-guo !

**Comment by lukepass** [comment]
Hello and thanks for the excellent job! Do I have to specify a special branch in my composer.json to get this feature?

**Comment by greg0ire** [comment]
Sounds as it you want a release, let me create one!

**Comment by lukepass** [comment]
Thank you very much, all my tests are OK after the update!

**Comment by lukepass** [comment]
I just noticed that with this new version, the following code breaks if the media is null:

```
{% thumbnail value, 'admin' %}
```

This is also used in the list_image.html.twig template:

```
{#

This file is part of the Sonata package.

(c) Thomas Rabaix <thomas.rabaix@sonata-project.org>

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.

#}

{% extends '@SonataAdmin/CRUD/base_list_field.html.twig' %}

{% block field%}
    <a href="{{ admin.generateUrl('edit', {'id' : object|sonata_urlsafeid }) }}">{% thumbnail attribute(object, field_description.name ), 'admin' with {'width': 75, 'height': 60} %}</a>
{% endblock %}
```

Error:
```
Type error: strlen() expects parameter 1 to be string, null given
```

**Comment by OskarStark** [comment]
Can you please open a new issue?

**Comment by lukepass** [comment]
Yes, here you are: https://github.com/sonata-project/SonataMediaBundle/issues/1517
I think ut's quite critical because it breaks all the page using null medias and I had to revert to the previous version.

**Pull request closed without merging**